src/ConfigSpace/hyperparameters/_hp_components.py

```diff
@@ -0,0 +1,337 @@
+from __future__ import annotations
```

Comment:
The reason for the abstraction of things in this file is that a lot of methods for the hyperparameters were essentially doing the same things, i.e. sampling from some vectorized distribution and then transforming to and from.

The original thought was that this would be one of the cythonized files, i.e. allow the hyperparameter files to be seen in editors and see their definitions and workings, however this also ended up with lifting all behavior to the base Hyperparameter class. This helps for a few reasons, namely:
1. A single source for behaviour of hyperparameters, reducing duplication/inconsistencies.
2. Implementing hyperparameters is mostly just passing the different components described above to the base `Hyperparameter` class at this point. You can see this from the individual hyperparameter classes that exist.
3. Any optimizations done for the underlying shared components can apply to all hyperparmaters, for example neighbor generation or sampling.

It's not perfect of course but really it's trying to make hyperparameters rely more on compositional structure rather than hierarchical behaviors.